website/addons/s3/static/tests/s3NodeConfig.test.js

```diff
@@ -0,0 +1,435 @@
+            }, {
+                description: 'Node is authorized and User is auth owner',
+                endpoint: makeSettingsEndpoint({
+                    node_has_auth: true,
+                    user_has_auth: true,
+                    user_is_owner: true,
+                    owner: faker.name.findName(),
+                    bucket: null
+                }),
+                data: {
+                    showSettings: true,
+                    disableSettings: false,
+                    showNewBucket: true,
+                    showImportAuth: false,
+                    showCreateCredentials: false,
+                    canChange: true,
+                    allowSelectBucket: false
+                }
+            }]);
+    });
+    describe('#toggleBucket', () => {
+        var server;
+        var endpoints = [{
+                method: 'GET',
+                url: URLS.bucket_list,
+                response: {
+                    buckets: new Array(10).map(faker.internet.password)
+                }
+            },
+            makeSettingsEndpoint()
+        ];
+        before(() => {
+            server = utils.createServer(sinon, endpoints);
+        });
+        after(() => {
+            server.restore();
+        });
+        it('shows the bucket selector when disabled and if buckets aren\'t loaded fetches the list of buckets', (done) => {
+            var vm = new s3NodeConfigVM('/api/v1/12345/s3/settings/', '', '/12345');
+            vm.fetchFromServer()
+                .always(function() {
+                    vm.showSelect(false);
+                    vm.loadedBucketList(false);
+                    var spy = sinon.spy(vm, 'fetchBucketList');
+                    var promise = vm.toggleSelect();
+                    promise.always(function() {
+                        assert.isTrue(vm.showSelect());
+                        assert.isTrue(vm.loadedBucketList());
+                        assert.isAbove(vm.bucketList().length, 0);
+                        assert(spy.calledOnce);
+                        done();
+                    });
+                });
+        });
+    });
+    describe('#selectBucket', () => {
+        var postEndpoint = makeSettingsEndpoint();
+        postEndpoint.method = 'POST';
+        postEndpoint.response = postEndpoint.response.result;
+        var bucket = faker.internet.domainWord();
+        postEndpoint.response.bucket = bucket;
+        postEndpoint.response.has_bucket = true;
+        var endpoints = [
+            postEndpoint,
+            makeSettingsEndpoint()
+        ];
+        var server;
+        before(() => {
+            server = utils.createServer(sinon, endpoints);
+        });
+        after(() => {
+            server.restore();
+        });
+        it('submits the selected bucket to the server, and updates data on success', (done) => {
+            var vm = new s3NodeConfigVM('/api/v1/12345/s3/settings/', '', '/12345');
+            vm.fetchFromServer()
```

Comment:
@sloria  For purity's sake maybe the best way is to make the synchronous call to updateFromData  instead of relying on fetchFromServer to initialize the VM?